tools/topology/topology2/platform/intel/sdw-amp-feedback.conf

```diff
@@ -0,0 +1,64 @@
+Object.Dai {
+	ALH."515" {
```

Comment:
I know I asked this before, but I don't recall what the answer was

Can we really encode this '515' value which is related to ALH channels? this might depend on which link and ALH channel is used for the feedback, no? So if this is in an include file should this be a parameter?

Reply:
> I know I asked this before, but I don't recall what the answer was
> 
> Can we really encode this '515' value which is related to ALH channels? this might depend on which link and ALH channel is used for the feedback, no? So if this is in an include file should this be a parameter?

@plbossart The ALH index (`515` here) is not related to ALH channels. We get `alh_stream_id` from [sdw_params_stream](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/intel/hda.c#L92) and set [copier_data->gtw_cfg.node_id |= SOF_IPC4_NODE_INDEX(data->dai_data);](https://github.com/thesofproject/linux/blob/topic/sof-dev/sound/soc/sof/ipc4-topology.c#L2330)
The only restriction of ALH index is that we can't have duplicated ALH index in a topology. In other words, we can have ALH.0, ALH.1, ALH.2, ALH.3, ALH.4 if we want.

Reply:
oh, so if this is a convention that's totally arbitrary we should not use 515 which was from another era and tied to a specific link/PDI usage. Should we use the same numbers as for PCM devices then in our ALH. numbering? I mean the speaker could be ALH.2 and mic ALH.4.

Reply:
> oh, so if this is a convention that's totally arbitrary we should not use 515 which was from another era and tied to a specific link/PDI usage. Should we use the same numbers as for PCM devices then in our ALH. numbering? I mean the speaker could be ALH.2 and mic ALH.4.

I choose the same ALH. number as for IPC3 ALH number is because that way can guarantee we will not use duplicated ALH. number. A problem of using the same number as for PCM devices is that we need two ALH in an aggregated speaker device. For example, currently we have ALH.258 and ALH.514 for one aggregated speaker device, and we can't use ALH.2-1 and ALH.2-2

Reply:
> the ALH number is not fixed for IPC3, it's based on macros. I don't mind if we use the same convention but we don't want to encode the link position in the topology, because we'll have to redo this when we support different hardware layouts.

@plbossart Yeah, I know the ALH number is not fixed for IPC3. I meant in the most common case, we have

SDW0 DP2 for headset playback
SDW0 DP3 for headset capture
SDW1 DP2 for 1st speaker playback
SDW2 DP2 for 2nd speaker playback
SDW3 DP3 for dmic capture

Based on above layout I assigned IPC4 ALH number with the same formula as IPC3. i.e. 0x002 for headset playback, 0x003 for headset capture, 0x102 for 1st speaker, etc.

But we don't really care about the real HW layouts. We can still use ALH.2 for headset playback if the headset codec is not on SDW0.
In other words, the ALH index number doesn't need to be changed when the HW layouts are changed.
And we can use the same convention to add a new ALH dai.

Does it make sense?

Reply:
sorry, I am not following what IPC3 convention you are referring to @bardliao since we don't have one? 

Or maybe you are referring to the initial CML designs with 1 and 2 amps?

Reply:
> sorry, I am not following what IPC3 convention you are referring to @bardliao since we don't have one?
> 
> Or maybe you are referring to the initial CML designs with 1 and 2 amps?

@plbossart I was referring to the (link_id * 256 + data port) formula. But I just realize that we can use the same rule as the pipeline id and the route id. I.e. pcm id * 10 ~ pcm id * 10 + 9.
I will update the PR to follow the rule.